### PR TITLE
Avoid duplicate dir search

### PR DESCRIPTION
**Pull request by vdukhovni**
This avoids adding the default CApath to the list of sources for "store" lookups, it is already on the legacy directory search list.

Is including the default search directory (or else the `SSL_CERT_DIR` environment variable value) in the store lookup list required in some situations???

Note, this is based on PR #24139 that fixes tests that failed (depending on time of day), while testing this PR.

### CODE REVIEW

**Review by nhorman — APPROVED** [review]

File: crypto/x509/x509_d2.c
Comment (vdukhovni) [reply crypto/x509/x509_d2.c:33]: If we some day have a default URI, presumably we'd add a call to instantiate that, rather than this call?
Though, yes, it is a bit odd that the default directory is silently added when a NULL source is supplied, that case should be a NOP, or maybe even return an error indication about the missing argument.

[ I did read the code you linked to by the way, and decided to leave it alone for now, in case some existing code, perhaps even outside OpenSSL, relies on its current behaviour. ]

Comment (levitte) [reply crypto/x509/x509_d2.c:33]: > If we some day have a default URI, presumably we'd add a call to instantiate that, rather than this call?

Er, isn't this function the generic call to instantiate all defaults?  Why would a STORE URI be excluded?

But otherwise, there's a reason I'm of two minds here...  STORE URIs have the `file:` scheme as implicit default, i.e. you can give it any file path and it is able to use that (by loading an implementation of the `file:` scheme, which we have built in).  So conceptually speaking, a directory *is a* URI and is perfectly usable with the STORE API.

However, thinking about this further, it may be questionable to use the *built in default directory* as a built in default URI, so all things considered, I'm still convinced that changing `crypto/x509/by_store.c` is a better move...  and at least for the moment, I'd say that the `X509_L_ADD_STORE` handler could simply return 1 if `argp == NULL`.

Comment (vdukhovni) [reply crypto/x509/x509_d2.c:33]: I had actually considered the option you suggested, but was hesitant to go there at the time. The result is I think cleaner (no kluky magic with the default directory). So I'm happy you see that as the right way forward. The call I deleted now becomes a NOP, but perhaps some day some other (non-duplicative) default might result from that call?

If calls with argp == NULL will always just short-circuit to NOP, why make them?  That is, after simplifying the callee, should *still* remove the call, or if not, why not?

Comment (levitte) [reply crypto/x509/x509_d2.c:33]: I would keep that call, because like you say, there may be a default that gets inserted into `crypto/x509/by_store.c` some day...  will we remember to add back the line this PR removes?  History shows that we're [ahem] collectively a bit prone to forget that sort of thing...

Comment (vdukhovni) [reply crypto/x509/x509_d2.c]: Done.

**Comment by vdukhovni** [comment]
Performance comparison from `x509storeissuer`, with thread counts of 1, 10, 100 and 1000:

Baseline is master without this PR, 

|Threads|Baseline (us)|PR (us)|
|---|---|---|
|   1 | 3107|1459|
|  10 | 8505|2432
| 100 | 156688|2560|
|1000 |1382203|2501|

**Comment by vdukhovni** [comment]
I think this is ready for review (see updated comment above). Dropping legacy engine support from the store is I think a separate issue, that can be tackled in another PR... @levitte perhaps?

**Review by Sashan — APPROVED** [review]
thanks for fixing the tests. I think I've hit this bug earlier this week when running tests late in night my time.

**Comment by vdukhovni** [comment]
> thanks for fixing the tests. I think I've hit this bug earlier this week when running tests late in night my time.

Note that the test fixes are a [separate PR](https://github.com/openssl/openssl/pull/24139) that is a merge-base for this one. Once that is merged, this will be rebased to contain only the last commit. The base PR has two approvals, so I hope will be merged shortly.

**Comment by levitte** [comment]
> I think this is ready for review (see updated comment above). Dropping legacy engine support from the store is I think a separate issue, that can be tackled in another PR... @levitte perhaps?

That would be a breaking change for anyone still using an engine with a scheme implementation.  I believe that the pkcs11 engine is one of those...  no?  Anyway, that would defer such a change to OpenSSL 4.0 at the earliest, and should probably happen in sync with general removal of engine support (which is currently only deprecated).

**Comment by levitte** [comment]
> IMHO the URI schemes should still be **eagerly** resolved to their lookup methods at the time the URI is added to the store, rather than during the lookup. Are there compelling reasons to defer the binding?

This sounds like a "what function to use" question, really.  We have two functions to add a URI to a OSSL_STORE lookup:

- `X509_LOOKUP_add_store()`, which just adds the URI to an internal list, from which objects will be loaded later.  This is the lazy function.
- `X509_LOOKUP_load_store_ex()` (and the older `X509_LOOKUP_load_store()`), which loads all objects from the given URI immediately.  This is the eager function.

So it seems to me that the choice is left to the user.

But, perhaps you're trying to say that we should use `X509_LOOKUP_load_store_ex()` in `X509_STORE_set_default_paths_ex()`, rather than `X509_LOOKUP_add_store()`?

**Comment by t8m** [comment]
Please rebase as the test fix PR was merged.

File: crypto/x509/by_store.c
Comment (mattcaswell) [inline crypto/x509/by_store.c:114]: So is this a change in behaviour in the case that someone *only* adds the store lookup? i.e. could it break someone relying on this?

Comment (vdukhovni) [reply crypto/x509/by_store.c:114]: Yes, they'd have to add some actual URIs to get anything to happen. I believe that's the right behaviour, but yes, it is different. One might for example ask why only the CApath and not the CAfile defaults into the store...

**Comment by vdukhovni** [comment]
Nudging reviewers... I believe this is ready.

File: crypto/x509/x509_d2.c
Comment (vdukhovni) [reply crypto/x509/x509_d2.c]: Reworked and moved the comment.

**Review by t8m — APPROVED** [review]
@nhorman please reconfirm

**Comment by nhorman** [comment]
reconfirmed, approval holds

**Comment by t8m** [comment]
Merged to the master branch. Thank you.

**Pull request closed without merging**
